public/app/core/components/TimelineChart/timeline.ts

```diff
@@ -89,6 +91,17 @@ export function getConfig(opts: TimelineCoreOptions) {
       .map((v) => Array(count).fill(null));
   };
 
+  // handle null and NaN mapped values when drawing a box
+  const shouldDrawYVal = (sidx: number, yVal: number | null): boolean => {
+    if (yVal == null) {
+      return isDiscrete(sidx) && hasMappedNull(sidx);
+    }
+    if (Number.isNaN(yVal)) {
+      return isDiscrete(sidx) && hasMappedNaN(sidx);
+    }
+    return true;
+  };
+
```

Comment:
this one is going to be much more expensive to run.

`isDiscrete(sidx)`, hasMappedNaN(sidx), and `hasMappedNull(sidx)` are invariant for the entire hot loop where this fn is used. there can easily by 2000 values x 30 series rendered at the same time, so this needs to be extremely efficient and not do unnecessary checks for every value. let's restructure to avoid this plz :pray:.

Reply:
Thanks for this performance related insight @leeoniya. 🚀 Good catch WRT the invariant + hot loop, I'll take a look and see how we can fine tune this a little more.